indico/client/request.py

```diff
@@ -1,8 +1,16 @@
+from abc import abstractmethod
 from enum import Enum
-from typing import Any, Dict, Union
+from typing import TYPE_CHECKING, Generic, TypeVar, cast
 
 from indico.errors import IndicoRequestError
 
+if TYPE_CHECKING:  # pragma: no cover
+    from typing import Any, Dict, Iterator, List, Optional, Union
+
+    from indico.typing import AnyDict
+
+ResponseType = TypeVar("ResponseType")
```

Comment:
note: should this be `covariant=True`? that would enable stuff like this
```py
workflow: Workflow | None = None
workflow = client.call(GetWorkflow(...))
```
which is currently disallowed

not super clear to me if there are any negative repercussions of this

Reply:
Personally worked around this by not using None and raising an error if a workflow is expected. I find try except pattern to cover my cases usually.

Is this a new limitation in python 3.12? I don't think I've had this typing issue before.

Reply:
not new, mypy 1.8 on Python 3.10

covariance also allows this, which I could see as more of an argument as to why it's useful
```py
datasets: list[Dataset] = []
datasets.extend(
  client.call(ListDatasets(...))
)
```
because std `extend` is typed as
```py
def extend(self, iterable: Iterable[_T])
```
(so Iterable vs. List)